File: scripts/diagram_updater.py
import os
import json
import time
from threading import Thread
from adjacency_list import AdjacencyList
from visualiser import Visualiser


class DiagramUpdater:
    def __init__(self, json_file_path, output_file_path, direction="TD", poll_interval=2):
        """
        Initialize the DiagramUpdater.

        :param json_file_path: Path to the JSON file to monitor.
        :param output_file_path: Path to the output Mermaid diagram file.
        :param direction: Direction of the Mermaid diagram (e.g., "TD", "LR").
        :param poll_interval: Time interval (in seconds) to poll for changes.
        """
        self.json_file_path = json_file_path
        self.output_file_path = output_file_path
        self.direction = direction
        self.poll_interval = poll_interval
        self.last_modified_time = None
        self.running = False
        self.adjacency_list = AdjacencyList()
        self.visualiser = Visualiser()

    def check_for_changes(self):
        """
        Check if the JSON file has been modified since the last check.
        """
        try:
            current_modified_time = os.path.getmtime(self.json_file_path)
            if self.last_modified_time is None or current_modified_time > self.last_modified_time:
                self.last_modified_time = current_modified_time
                return True
        except FileNotFoundError:
            print(f"ERROR: JSON file not found: {self.json_file_path}")
        return False

    def update_diagram(self):
        """
        Generate the Mermaid diagram from the JSON file.
        """
        try:
            # Read and parse the JSON file
            with open(self.json_file_path, 'r') as f:
                component_dumps = json.load(f)

            # Build the adjacency list
            adjacency_list = self.adjacency_list.build_adjacency_list(component_dumps)

            # Generate the Mermaid diagram
            mermaid_content = self.visualiser.build_mermaid_diagram(adjacency_list, self.direction)

            # Write the Mermaid diagram to the output file
            os.makedirs(os.path.dirname(self.output_file_path), exist_ok=True)
            with open(self.output_file_path, 'w') as f:
                f.write(mermaid_content)

            print(f"Updated Mermaid diagram at: {self.output_file_path}")
        except Exception as e:
            print(f"ERROR: Failed to update Mermaid diagram: {str(e)}")

    def start(self):
        """
        Start the DiagramUpdater session.
        """
        if self.running:
            print("DiagramUpdater is already running.")
            return

        self.running = True
        print(f"Starting DiagramUpdater for {self.json_file_path}...")

        def run():
            while self.running:
                if self.check_for_changes():
                    print(f"Changes detected in {self.json_file_path}. Updating diagram...")
                    self.update_diagram()
                time.sleep(self.poll_interval)

        Thread(target=run, daemon=True).start()

    def stop(self):
        """
        Stop the DiagramUpdater session.
        """
        self.running = False
        print("DiagramUpdater stopped.")


# Example usage
if __name__ == "__main__":
    json_file = "demo-jvm/build/knit.json"
    output_file = "demo-jvm/build/knit_diagram.mmd"

    updater = DiagramUpdater(json_file, output_file, direction="TD", poll_interval=2)
    try:
        updater.start()
        while True:
            time.sleep(1)  # Keep the main thread alive
    except KeyboardInterrupt:
        updater.stop()

File: scripts/DiagramUpdater.ts
// DiagramUpdater.ts

import { AdjacencyList } from './AdjacencyList';
//import { Visualiser } from './Visualiser';
import { IncrementalAdjacency } from './IncrementalAdjacency';

export class DiagramUpdater {
  jsonFileUrl: string;
  outputCallback: (content: string) => void;
  direction: string;
  pollInterval: number;
  lastModifiedTime: number | null = null;
  changesDirUrl: string | null = null;
  lastChangeSeen: string | null = null;
  running: boolean = false;
  adjacencyList: AdjacencyList;
  //visualiser: Visualiser;
  incremental: IncrementalAdjacency;
  currentAdj: Record<string, string[]> | null = null;
  intervalId: number | null = null;

  constructor(
    jsonFileUrl: string,
    outputCallback: (content: string) => void,
    direction = "TD",
    pollInterval = 2000,
    changesDirUrl?: string
  ) {
    this.jsonFileUrl = jsonFileUrl;
    this.outputCallback = outputCallback;
    this.direction = direction;
    this.pollInterval = pollInterval;
    this.changesDirUrl = changesDirUrl || null;
    this.adjacencyList = new AdjacencyList();
    //this.visualiser = new Visualiser();
    this.incremental = new IncrementalAdjacency();
  }

  async fetchJson(url: string): Promise<any> {
    const response = await fetch(url, { cache: "no-store" });
    if (!response.ok) throw new Error(`Failed to fetch ${url}`);
    return await response.json();
  }

  async checkForChanges(): Promise<boolean> {
    // In browser, we can't get mtime, so we refetch and compare content hash
    try {
      const data = await this.fetchJson(this.jsonFileUrl);
      const hash = JSON.stringify(data);
      if (this.lastModifiedTime === null || hash !== String(this.lastModifiedTime)) {
        this.lastModifiedTime = hash as any;
        return true;
      }
    } catch (e) {
      console.error(`ERROR: JSON file not found: ${this.jsonFileUrl}`);
    }
    return false;
  }

  async updateDiagramFull(): Promise<void> {
    try {
      const componentDumps = await this.fetchJson(this.jsonFileUrl);
      const adjacencyList = this.adjacencyList.buildAdjacencyList(componentDumps);
      this.currentAdj = adjacencyList;
      this.adjacencyList.resetLastUpdateFlags();
      this.adjacencyList.markFullBuildUpdate();
      this.adjacencyList.propagateErrorFlagsFromAdj(this.currentAdj);
      //const mermaidContent = this.visualiser.buildMermaidDiagram(adjacencyList, this.direction);
      //this.outputCallback(mermaidContent);
      console.log(`Updated diagram`);
    } catch (e) {
      console.error(`ERROR: Failed to update diagram: ${e}`);
    }
  }

  async latestChangeFile(): Promise<string | null> {
    if (!this.changesDirUrl) return null;
    try {
      const latestTxtUrl = this.changesDirUrl + "/latest.txt";
      const response = await fetch(latestTxtUrl, { cache: "no-store" });
      if (!response.ok) return null;
      const name = (await response.text()).trim();
      const changePath = this.changesDirUrl + "/" + name;
      const changeResp = await fetch(changePath, { cache: "no-store" });
      if (!changeResp.ok) return null;
      return changePath;
    } catch {
      return null;
    }
  }

  async checkForChangeFiles(): Promise<string | null> {
    const changePath = await this.latestChangeFile();
    if (!changePath) return null;
    if (changePath !== this.lastChangeSeen) {
      this.lastChangeSeen = changePath;
      return changePath;
    }
    return null;
  }

  async applyIncrementalIfAny(): Promise<boolean> {
    if (!this.currentAdj) return false;
    const changeFile = await this.checkForChangeFiles();
    if (!changeFile) return false;
    try {
      const changeData = await this.fetchJson(changeFile);
      this.adjacencyList.resetLastUpdateFlags();
      this.adjacencyList.applyStatusChange(changeData);
      this.currentAdj = this.incremental.applyChange(this.currentAdj, changeData);
      this.adjacencyList.propagateErrorFlagsFromAdj(this.currentAdj);
      //const mermaidContent = this.visualiser.buildMermaidDiagram(this.currentAdj, this.direction);
      //this.outputCallback(mermaidContent);
      const nodeCount = Object.keys(this.currentAdj).length;
      const edgeCount = Object.values(this.currentAdj).reduce((acc, v) => acc + v.length, 0);
      console.log(`Applied change -> nodes=${nodeCount}, edges=${edgeCount}; updated diagram.`);
      return true;
    } catch (e) {
      console.error(`ERROR: Failed to apply change file: ${e}`);
      return false;
    }
  }

  async start(): Promise<void> {
    if (this.running) {
      console.log("DiagramUpdater is already running.");
      return;
    }
    this.running = true;
    console.log(`Starting DiagramUpdater for ${this.jsonFileUrl}...`);
    await this.updateDiagramFull();

    this.intervalId = window.setInterval(async () => {
      const applied = await this.applyIncrementalIfAny();
      if (await this.checkForChanges() && !applied) {
        console.log(`Changes detected in ${this.jsonFileUrl}. Updating diagram...`);
        await this.updateDiagramFull();
      }
    }, this.pollInterval);
  }

  stop(): void {
    this.running = false;
    if (this.intervalId !== null) window.clearInterval(this.intervalId);
    console.log("DiagramUpdater stopped.");
  }
}
